Libraries/Connect/Implementation/Interceptors/InterceptorChain.swift

```diff
@@ -14,79 +14,79 @@
 /// and orchestrates invoking each of them in the proper order.
-struct InterceptorChain: @unchecked Sendable {
-    private let interceptors: [Interceptor]
+final class InterceptorChain<T: Sendable>: Sendable {
+    private let interceptors: [T]
 
-    /// Initialize the interceptor chain.
-    ///
-    /// NOTE: Exactly 1 chain is expected to be instantiated for a single request/stream.
-    ///
-    /// - parameter interceptors: Closures that should be called to create interceptors.
-    /// - parameter config: Config to use for setting up interceptors.
-    init(interceptors: [InterceptorInitializer], config: ProtocolClientConfig) {
-        self.interceptors = interceptors.map { initialize in initialize(config) }
+    init(_ interceptors: [T]) {
+        self.interceptors = interceptors
     }
 
-    /// Create a set of closures configured with all interceptors for a unary API.
+    /// Invoke each of the interceptors, waiting for a given interceptor to complete before passing
+    /// the resulting value to the next interceptor and finally invoking `finish` with the final
+    /// value.
     ///
-    /// NOTE: Interceptors are invoked in FIFO order for the request path, and in LIFO order for
-    /// the response path. For example, with interceptors `[a, b, c]`:
-    /// `caller -> a -> b -> c -> server`
-    /// `caller <- c <- b <- a <- server`
-    ///
-    /// - returns: A set of closures that each invoke the chain of interceptors in the above order.
-    func unaryFunction() -> UnaryFunction {
-        let interceptors = self.interceptors.map { $0.unaryFunction() }
-        return UnaryFunction(
-            requestFunction: { request in
-                return executeInterceptors(interceptors.map(\.requestFunction), initial: request)
-            },
-            responseFunction: { response in
-                return executeInterceptors(
-                    interceptors.reversed().map(\.responseFunction),
-                    initial: response
-                )
-            },
-            responseMetricsFunction: { metrics in
-                return executeInterceptors(
-                    interceptors.reversed().map(\.responseMetricsFunction),
-                    initial: metrics
-                )
-            }
-        )
+    /// - parameter functionPath: Key path of the function on the interceptor to invoke.
+    /// - parameter firstInFirstOut: If true, interceptors will be invoked in the order they were
+    ///                              originally registered. If false, the order will be reversed.
+    /// - parameter initial: The initial value to pass to the first interceptor.
+    /// - parameter finish: Closure to call with the final value after each interceptor has finished
+    ///                     processing.
+    func executeInterceptors<Value>(
+        _ functionPath: KeyPath<T,
+            @Sendable (Value, @escaping @Sendable (Value) -> Void) -> Void
+        >,
```

Comment:
Newlines here look a bit off - perhaps:

```suggestion
        _ functionPath: KeyPath<T, @Sendable (Value, @escaping @Sendable (Value) -> Void) -> Void>,
```

Reply:
Updated